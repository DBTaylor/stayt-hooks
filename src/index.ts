import {Stayt} from 'stayt'
import {useState, useEffect, useMemo, memo} from 'react'

export const useStayt = <T>(state: Stayt<T>) => {
    const [value, setValue] = useState(() => state.get())
    useEffect(() => {
        const v = state.get()
        if(v !== value)
            setValue(v)
        const fn = (val: T) => setValue(val)
        state.subscribe(fn)
        return () => state.unsubscribe(fn)
    }, [state])
    return value
}

export const useOptionStayt = <T>(state: Stayt<T | null>) => {
    const [isNull, setIsNull] = useState(() => state.get() === null)
    const option = useMemo(() => state.option(), [state])
    useEffect(() => {
        const isNull2 = state.get() === null
        if(isNull2 !== isNull)
            setIsNull(isNull2)
        const fn = (n: boolean) => setIsNull(n)
        state.subscribeNull(fn)
        return () => state.unsubscribeNull(fn)
    }, [state])
    return isNull ? null : option
}

type Narrow<T, N> = T extends { kind: N } ? T : never;

// export const useOptionStayt = <T>(state: Stayt<T>) => {
//     const value = useStayt(state)
//     const option = useMemo(() => state.option(), [state])
//     if(value !== null)
//         return option
//     else
//         return null
// }

type MapDisc<T, U extends T[keyof T & "kind"]> = U extends unknown ? {kind: U, state: Stayt<Narrow<T,U>>} : undefined

export const useDiscStayt = <U extends T["kind" & keyof T] & string, T extends {kind: string}>(state: Stayt<T>) => {
    const kindStayt = useMemo(() => state.prop("kind"), [state])
    const kind = useStayt(kindStayt)
    return useMemo(() => ({kind, state: state.disc(kind as any)} as any as MapDisc<T, U>), [kind])
}

export const useArrStayt = <T>(state: Stayt<T[]>) => {
    const len = useStayt(state.prop("length"))
    return useMemo(() =>{
        const arr = Array(len)
        for (let i = 0; i < len; ++i) {
            arr[i] = state.index(i);
        }
        return arr
    }, [len])
}