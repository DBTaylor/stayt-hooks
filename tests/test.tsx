import {Stayt} from 'stayt'
import {useStayt, useOptionStayt, useDiscStayt, useArrStayt} from '../src/index'
import renderer from 'react-test-renderer';
import React from 'react'
const {useState} = React
const {act} = renderer

{
    type T = {kind: "string", value: string} | {kind: "number", value: number}
    let i = 0;
    let j = 0;
    let k = 0;
    const Num = ({v}: {v: Stayt<number>}) => {
        i++
        const n = useStayt(v)
        return <p>{i} {n + n}</p>
    }
    const Str = ({v}: {v: Stayt<string>}) => {
        const s = useStayt(v)
        j++
        return <p>{j} {s}</p>
    }
    const Test = ({v}: {v: Stayt<T>}) => {
        k++
        const val = useDiscStayt(v)
        return <>
            <p>{k}</p>
            {val.kind === "number" ?
                <Num v={val.state.prop("value")}/>
            :
                <Str v={val.state.prop("value")}/>
            }
        </>
    }
    const state: Stayt<T> = new Stayt({kind: "string", value: "Hello World"})
    test('Test', async () => {
        const component = renderer.create(
            <Test v={state}/>,
        );

        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()

        act(() => state.disc("string").prop("value").modify(s => s + "!"))
        tree = component.toJSON()

        expect(tree).toMatchSnapshot()
        act(() => state.modify(_ => ({kind: "number", value: 3})))

        tree = component.toJSON()
        expect(tree).toMatchSnapshot()

        act(() => state.disc("number").prop("value").modify(n => n + 1))
        tree = component.toJSON()

        expect(tree).toMatchSnapshot()
    })
}
{
    const Num = ({v}: {v: Stayt<number>}) => {
        const n = useStayt(v)
        return <p>{n + n}</p>
    }
    const Test = ({v}: {v: Stayt<number[]>}) => {
        const val = useArrStayt(v)
        return <>
            {val.map((n, i) => <Num key={i} v={n}/>)}
        </>
    }
    const state = new Stayt([1, 2, 3, 4])
    test('ArrTest', async () => {
        const component = renderer.create(
            <Test v={state}/>,
        );
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot()
        act(() => state.index(2).modify(n => n + 1))
        tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })

}