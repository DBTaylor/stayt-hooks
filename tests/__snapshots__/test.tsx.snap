// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrTest 1`] = `
Array [
  <p>
    2
  </p>,
  <p>
    4
  </p>,
  <p>
    6
  </p>,
  <p>
    8
  </p>,
]
`;

exports[`ArrTest 2`] = `
Array [
  <p>
    2
  </p>,
  <p>
    4
  </p>,
  <p>
    8
  </p>,
  <p>
    8
  </p>,
]
`;

exports[`OptTest 1`] = `<div />`;

exports[`OptTest 2`] = `
<p>
  10
</p>
`;

exports[`OptTest 3`] = `<div />`;

exports[`Test 1`] = `
Array [
  <p>
    1
  </p>,
  <p>
    1
     
    Hello World
  </p>,
]
`;

exports[`Test 2`] = `
Array [
  <p>
    1
  </p>,
  <p>
    2
     
    Hello World!
  </p>,
]
`;

exports[`Test 3`] = `
Array [
  <p>
    2
  </p>,
  <p>
    1
     
    6
  </p>,
]
`;

exports[`Test 4`] = `
Array [
  <p>
    2
  </p>,
  <p>
    2
     
    8
  </p>,
]
`;
